<a href="https://colab.research.google.com/github/oddecaL/Parkinson-s-Disease-Early-Detection/blob/main/EDA_Parkinsons.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Importing all the required functions




In [ ]:
import kagglehub
import os
import pandas as pd
import numpy as np

# Loading synthetic and actual datasets

## Synthetic Dataset

In [ ]:
syn_path = kagglehub.dataset_download(
    "miadul/parkinsons-disease-synthetic-dataset"
)

print(syn_path)

100%|██████████| 614k/614k [00:00<00:00, 991kB/s]

Extracting files...
/root/.cache/kagglehub/datasets/miadul/parkinsons-disease-synthetic-dataset/versions/1


In [ ]:
print(os.listdir(syn_path))

['parkinsons_disease_dataset.csv']


In [ ]:
syn_csv_path = os.path.join(syn_path, "parkinsons_disease_dataset.csv")

syn_df = pd.read_csv(syn_csv_path)

In [ ]:
syn_df

,Age,Sex,Tremor_Intensity,Voice_Pitch,Voice_Jitter,Voice_Shimmer,Bradykinesia_Score,Postural_Instability,Rigidity_Level,UPDRS_Score,Handwriting_Change,Walking_Difficulty,Family_History,MRI_Score,Parkinsons_Status
0,78,Male,5.19,287.62,3.82,2.98,8.05,9.11,6.67,79.13,0.97,7.39,No,11.41,1
1,68,Female,6.22,281.92,3.38,3.79,1.04,7.64,1.43,97.50,1.67,1.71,Yes,12.58,0
2,54,Female,4.20,231.59,2.89,4.03,8.14,2.05,3.80,73.40,6.30,9.79,Yes,1.55,0
3,82,Male,8.77,236.50,3.16,2.89,5.35,6.52,4.87,87.81,3.41,2.33,No,92.83,1
4,47,Male,5.23,100.48,4.60,3.43,6.80,8.13,3.59,94.33,6.41,0.52,Yes,36.37,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
22995,81,Female,4.10,102.12,0.72,3.93,7.18,9.45,9.81,46.06,5.65,9.90,Yes,36.29,0
22996,65,Male,0.25,290.78,0.12,2.29,0.32,2.43,6.39,8.81,3.94,5.09,No,16.44,0
22997,81,Female,4.19,254.47,4.97,3.70,4.79,0.79,6.68,23.80,9.71,4.38,No,53.74,0
22998,80,Male,6.60,239.11,2.31,1.60,4.90,8.03,6.94,35.20,9.72,9.04,No,82.09,1


In [ ]:
## Actual dataset

In [ ]:
path = kagglehub.dataset_download("gargmanas/parkinsonsdataset")

print("Path to dataset files:", path)

100%|██████████| 15.5k/15.5k [00:00<00:00, 15.6MB/s]

Extracting files...
Path to dataset files: /root/.cache/kagglehub/datasets/gargmanas/parkinsonsdataset/versions/1


In [ ]:
print(os.listdir(path))

['parkinsons.names', 'parkinsons.csv']


In [ ]:
actual_path = os.path.join(path,"parkinsons.csv")
df = pd.read_csv(actual_path)

In [ ]:
df

,name,MDVP:Fo(Hz),MDVP:Fhi(Hz),MDVP:Flo(Hz),MDVP:Jitter(%),MDVP:Jitter(Abs),MDVP:RAP,MDVP:PPQ,Jitter:DDP,MDVP:Shimmer,...,Shimmer:DDA,NHR,HNR,status,RPDE,DFA,spread1,spread2,D2,PPE
0,phon_R01_S01_1,119.992,157.302,74.997,0.00784,0.00007,0.00370,0.00554,0.01109,0.04374,...,0.06545,0.02211,21.033,1,0.414783,0.815285,-4.813031,0.266482,2.301442,0.284654
1,phon_R01_S01_2,122.400,148.650,113.819,0.00968,0.00008,0.00465,0.00696,0.01394,0.06134,...,0.09403,0.01929,19.085,1,0.458359,0.819521,-4.075192,0.335590,2.486855,0.368674
2,phon_R01_S01_3,116.682,131.111,111.555,0.01050,0.00009,0.00544,0.00781,0.01633,0.05233,...,0.08270,0.01309,20.651,1,0.429895,0.825288,-4.443179,0.311173,2.342259,0.332634
3,phon_R01_S01_4,116.676,137.871,111.366,0.00997,0.00009,0.00502,0.00698,0.01505,0.05492,...,0.08771,0.01353,20.644,1,0.434969,0.819235,-4.117501,0.334147,2.405554,0.368975
4,phon_R01_S01_5,116.014,141.781,110.655,0.01284,0.00011,0.00655,0.00908,0.01966,0.06425,...,0.10470,0.01767,19.649,1,0.417356,0.823484,-3.747787,0.234513,2.332180,0.410335
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
190,phon_R01_S50_2,174.188,230.978,94.261,0.00459,0.00003,0.00263,0.00259,0.00790,0.04087,...,0.07008,0.02764,19.517,0,0.448439,0.657899,-6.538586,0.121952,2.657476,0.133050
191,phon_R01_S50_3,209.516,253.017,89.488,0.00564,0.00003,0.00331,0.00292,0.00994,0.02751,...,0.04812,0.01810,19.147,0,0.431674,0.683244,-6.195325,0.129303,2.784312,0.168895
192,phon_R01_S50_4,174.688,240.005,74.287,0.01360,0.00008,0.00624,0.00564,0.01873,0.02308,...,0.03804,0.10715,17.883,0,0.407567,0.655683,-6.787197,0.158453,2.679772,0.131728
193,phon_R01_S50_5,198.764,396.961,74.904,0.00740,0.00004,0.00370,0.00390,0.01109,0.02296,...,0.03794,0.07223,19.020,0,0.451221,0.643956,-6.744577,0.207454,2.138608,0.123306


# Creating a user defined function for EDA Analysis

In [ ]:
def eda_summary(dt):
  # Data Shapes

  print("The shape of the dataset is ",dt.shape)
  print("\nThe number of rows of the dataset is ",dt.shape[0])
  print("\nThe number of columns of the dataset is ",dt.shape[1])

  # First 5 rows
  print("\nThe first 5 rows of the dataset is\n",dt.head())

  # Missing values
  missing = dt.isnull().sum()
  if missing.sum() == 0:
      print("\nNo missing values found.\n")
  else:
      print(missing[missing > 0])

  # Unique Values
  print("\nNo unique values ",dt.nunique())

  # Summary of the dataset
  print("\n",dt.describe())

# Performing Exploratory Data Analysis

## On the synthetic dataset

In [ ]:
print(eda_summary(syn_df))

NameError: name 'syn_df' is not defined

## On the Actual dataset

In [ ]:
print(eda_summary(df))

NameError: name 'df' is not defined

# **Dataset Overview**
The clinical dataset contains 195 records across 24 features — significantly smaller than the synthetic dataset. This small size is the primary motivation for using GANs to augment training data. No missing values were detected in the raw data, though the low unique value counts for some features (e.g., MDVP:Jitter(Abs) has only 19 unique values) suggest limited measurement resolution in those columns.

##**Feature Composition**
Features are exclusively biomedical voice measurements derived from sustained phonation recordings. They fall into three groups:
- **Frequency features**: MDVP:Fo(Hz), MDVP:Fhi(Hz), MDVP:Flo(Hz) — fundamental, maximum, and minimum vocal frequency
- **Jitter features** (frequency perturbation): MDVP:Jitter(%), MDVP:Jitter(Abs), MDVP:RAP, MDVP:PPQ, Jitter:DDP
- **Shimmer features** (amplitude perturbation): MDVP:Shimmer, MDVP:Shimmer(dB), Shimmer:APQ3, APQ5, MDVP:APQ, Shimmer:DDA
- **Nonlinear dynamics**: NHR, HNR, RPDE, DFA, spread1, spread2, D2, PPE

##**Fundamental Frequency (MDVP:Fo)**
The average vocal fundamental frequency ranges from 88.3 to 260.1 Hz with a mean of ~154.2 Hz. This is consistent with normal adult speech. The standard deviation of 41.4 Hz indicates considerable inter-subject variability, partly explained by the mix of male and female speakers in the dataset.

##**Jitter (Frequency Perturbation)**
MDVP:Jitter(%) has a mean of 0.62% and a maximum of 3.32%. Healthy individuals typically exhibit jitter below 1%, making elevated jitter a clinically meaningful indicator of vocal instability associated with Parkinson's. The right-skewed distribution (mean > median = 0.49%) suggests a subset of patients with significantly elevated jitter.

##**Shimmer (Amplitude Perturbation)**
MDVP:Shimmer has a mean of ~0.030 with a maximum of 0.119. Like jitter, elevated shimmer reflects neuromuscular control deterioration in the larynx — a hallmark of Parkinson's-related dysphonia. The distribution is also right-skewed.

##**Noise Ratios (NHR and HNR)**
NHR (Noise-to-Harmonics Ratio) has a mean of 0.025, while HNR (Harmonics-to-Noise Ratio) has a mean of 21.9 dB. Higher NHR and lower HNR both indicate increased breathiness and turbulence in voice — features strongly linked to Parkinson's vocal impairment.

##**Nonlinear Dynamics Features**
RPDE (mean 0.499) and DFA (mean 0.718) quantify the complexity and long-range correlation of vocal signals. spread1 (mean −5.68) and PPE (mean 0.207) measure fundamental frequency variation. These nonlinear features are among the most discriminative for PD detection, as Parkinson's disrupts the normal chaotic dynamics of healthy vocal production.

##**Target Variable (status)**
The mean status is ~0.754, meaning approximately 75.4% of the 195 records are Parkinson's positive (147 PD, 48 healthy) — a class imbalance of roughly 3:1. This is the core challenge the GAN addresses: the discriminator must learn to identify PD from a dataset where healthy examples are scarce.

##**Cardinality Observations**
Every continuous feature has high cardinality (148–195 unique values), consistent with real clinical measurements. The `name` column holds 195 unique patient identifiers and is dropped before modeling as it carries no predictive signal. The `status` column has exactly 2 unique values (binary classification target).

##**Key Implication**
With only 195 samples and a 3:1 class imbalance, a naive classifier could achieve ~75% accuracy by predicting Parkinson's for every sample. This makes accuracy an unreliable metric — Recall (sensitivity) is the primary metric of interest, ensuring the model minimises missed diagnoses.